vendor/k8s.io/kube-openapi/pkg/handler3/handler.go

```diff
@@ -207,16 +262,28 @@ func (o *OpenAPIService) RegisterOpenAPIV3VersionedService(servePath string, han
 	return nil
 }
 
-func (o *OpenAPIV3Group) UpdateSpec(specBytes []byte) (err error) {
+func (o *OpenAPIV3Group) UpdateSpec(openapi *spec3.OpenAPI) (err error) {
 	o.rwMutex.Lock()
 	defer o.rwMutex.Unlock()
 
+	o.jsonCache = o.jsonCache.New(func() ([]byte, error) {
+		return json.Marshal(openapi)
+	})
 	o.pbCache = o.pbCache.New(func() ([]byte, error) {
-		return ToV3ProtoBinary(specBytes)
+		json, err := o.jsonCache.Get()
+		if err != nil {
+			return nil, err
+		}
+		return ToV3ProtoBinary(json)
 	})
-
-	o.jsonCache = o.jsonCache.New(func() ([]byte, error) {
-		return specBytes, nil
+	// TODO: This forces a json marshal of corresponding group-versions.
+	// We should look to replace this with a faster hashing mechanism.
+	o.etagCache = o.etagCache.New(func() ([]byte, error) {
+		json, err := o.jsonCache.Get()
```

Comment:
is it correct to pin the etag to the json serialization (for the same etag to be returned to two different requests with two different body content)